driver-core/src/main/java/com/datastax/driver/core/QueryOptions.java

```diff
@@ -299,7 +301,19 @@ public QueryOptions setMetadataEnabled(boolean enabled) {
         boolean wasEnabled = this.metadataEnabled;
         this.metadataEnabled = enabled;
         if (!wasEnabled && enabled && manager != null) {
-            manager.submitSchemaRefresh(null, null, null, null); // will also refresh token map
+            // This is roughly the same as what we do in ControlConnection.tryConnect():
+            // 1. call submitNodeListRefresh() first to
+            // be able to compute the token map for the first time,
+            // which will be incomplete due to the lack of keyspace metadata
+            Futures.addCallback(manager.submitNodeListRefresh(), new MoreFutures.SuccessCallback<Void>() {
+                @Override
+                public void onSuccess(Void result) {
+                    // 2. then call submitSchemaRefresh() to
+                    // refresh schema metadata and re-compute the token map
+                    // this time with information about keyspaces
+                    manager.submitSchemaRefresh(null, null, null, null);
+                }
+            });
```

Comment:
The hosts list is kept up-to-date, but not the token map, which is likely to be `null` at this point. The field `Metadata.tokenMap` is only ever updated if `metadataEnabled == true` (see [here](https://github.com/datastax/java-driver/blob/3.0.x/driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java#L656-L657)), and the only way to build the token map from scratch for the first time is to trigger a node list refresh.